x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformFailureHandler.java

```diff
@@ -51,28 +52,32 @@ class TransformFailureHandler {
      */
-    void handleIndexerFailure(Exception e, SettingsConfig settingsConfig) {
+    void handleIndexerFailure(Exception exception, SettingsConfig settingsConfig) {
         // more detailed reporting in the handlers and below
-        logger.debug(() -> "[" + transformId + "] transform encountered an exception: ", e);
-        Throwable unwrappedException = ExceptionsHelper.findSearchExceptionRootCause(e);
+        logger.atDebug().withThrowable(exception).log("[{}] transform encountered an exception", transformId);
+        Throwable unwrappedException = ExceptionsHelper.findSearchExceptionRootCause(exception);
         boolean unattended = Boolean.TRUE.equals(settingsConfig.getUnattended());
 
-        if (unwrappedException instanceof CircuitBreakingException circuitBreakingException) {
-            handleCircuitBreakingException(circuitBreakingException, unattended);
-        } else if (unwrappedException instanceof ScriptException scriptException) {
-            handleScriptException(scriptException, unattended);
-        } else if (unwrappedException instanceof BulkIndexingException bulkIndexingException) {
-            handleBulkIndexingException(bulkIndexingException, unattended, getNumFailureRetries(settingsConfig));
-        } else if (unwrappedException instanceof ClusterBlockException clusterBlockException) {
+        if (unwrappedException instanceof CircuitBreakingException e) {
+            handleCircuitBreakingException(e, unattended);
+        } else if (unwrappedException instanceof ScriptException e) {
+            handleScriptException(e, unattended);
+        } else if (unwrappedException instanceof BulkIndexingException e) {
+            handleBulkIndexingException(e, unattended, getNumFailureRetries(settingsConfig));
+        } else if (unwrappedException instanceof ClusterBlockException e) {
             // gh#89802 always retry for a cluster block exception, because a cluster block should be temporary.
-            retry(clusterBlockException, clusterBlockException.getDetailedMessage(), unattended, getNumFailureRetries(settingsConfig));
-        } else if (unwrappedException instanceof ElasticsearchException elasticsearchException) {
-            handleElasticsearchException(elasticsearchException, unattended, getNumFailureRetries(settingsConfig));
-        } else if (unwrappedException instanceof IllegalArgumentException illegalArgumentException) {
-            handleIllegalArgumentException(illegalArgumentException, unattended);
+            retry(e, e.getDetailedMessage(), unattended, getNumFailureRetries(settingsConfig));
+        } else if (unwrappedException instanceof SearchPhaseExecutionException e) {
+            // The reason of a SearchPhaseExecutionException unfortunately contains a full stack trace.
+            // Instead of displaying that to the user, get the cause's message instead.
+            retry(e, e.getCause().getMessage(), unattended, getNumFailureRetries(settingsConfig));
```

Comment:
Will the cause always be non-null?

Reply:
I was just about to say that it cannot be `null` because that's enforced here: https://github.com/elastic/elasticsearch/blob/12ad399c488f0cc60e19b5e1b29c6d569cb4351a/server/src/main/java/org/elasticsearch/action/ShardOperationFailedException.java#L39

However, then I noticed that the exception we're treating specially here is `SearchPhaseExecutionException`. Should it be `ShardOperationFailedException` instead? That would be consistent with what was originally changed in #102200. But there are so many layers of wrapping and nesting that maybe by the time it gets here it's wrapped in a `SearchPhaseExecutionException`.

Reply:
There are 18 places in #45878 where `ExceptionsHelper.stackTrace()` is used. All of the ones that might affect transforms relate to searching shards.

If it's unclear whether to use `SearchPhaseExecutionException` or `ShardOperationFailedException` here then the safest thing might be to handle both, but account for the possibility that `e.getCause()` could be `null`.

Reply:
If you look at the [Github issue](https://github.com/elastic/ml-team/issues/1057), the exception message starts with
```
Failed to execute phase [
```
which indicates a `SearchPhaseExecutionException`.

The message contains submessages of the shards that failed (each between curly brackets). These nested `ShardSearchFailure`s contain the stack traces.

Reply:
Good catch about the non-null. 

Looking at the code I'd say it's non-null, because the `SearchPhaseExecutionException`'s implementation of `getCause` processes the nested `ShardOperationFailedException`s, which have non-null causes. But the code is non-trivial, I'm not 100% sure.

Reply:
I've added a null check, just in case